# Data Analytics Módulo 3
------------------------

## Ejercicio de evaluación final

#### Fase 1: Exploración y limpieza de datos

In [1]:
## Importamos las librerías que necesitaremos para la fase 1:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)

In [2]:
## Extraemos los ficheros:
df_vuelos_info = pd.read_csv("./files/Customer Flight Activity.csv")
df_clientes = pd.read_csv("./files/Customer Loyalty History.csv")

In [3]:
## Comprobamos el contenido del primer fichero:
df_vuelos_info.head()

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
0          100018  2017      1               3                        0   
1          100102  2017      1              10                        4   
2          100140  2017      1               6                        0   
3          100214  2017      1               0                        0   
4          100272  2017      1               0                        0   

   Total Flights  Distance  Points Accumulated  Points Redeemed  \
0              3      1521               152.0                0   
1             14      2030               203.0                0   
2              6      1200               120.0                0   
3              0         0                 0.0                0   
4              0         0                 0.0                0   

   Dollar Cost Points Redeemed  
0                            0  
1                            0  
2                            0  
3                            0  
4                            0

In [4]:
## Hacemos algunas comprobaciones:
print(df_vuelos_info.shape)
print("-----------------")
print(len(df_vuelos_info["Loyalty Number"].unique()))
print("-----------------")
print(df_vuelos_info ["Year"].unique())
print("-----------------")
print(df_vuelos_info["Loyalty Number"].unique().sum())
print("-----------------")
print(df_vuelos_info ["Points Redeemed"].unique())
print("-----------------")
print(df_vuelos_info ["Dollar Cost Points Redeemed"].unique())

(405624, 10)
-----------------
16737
-----------------
[2017 2018]
-----------------
9200929431
-----------------
[  0 341 364 310 445 312 343 366 389 292 447 324 456 409 436 327 322 291
 323 300 290 309 325 386 321 363 340 670 443 517 444 328 344 367 313 333
 293 449 297 455 372 356 405 381 466 419 369 352 482 335 329 305 415 396
 317 348 314 334 350 330 318 298 420 336 471 680 441 353 484 301 374 417
 501 299 398 307 368 306 347 439 395 481 337 382 426 373 399 424 326 392
 438 467 480 448 308 400 376 375 460 339 385 611 431 320 362 404 442 410
 361 319 435 414 464 477 315 485 370 421 349 371 416 496 510 667 465 434
 346 487 408 500 360 378 345 358 479 380 411 491 505 446 425 476 393 418
 332 401 454 303 594 506 355 302 403 379 437 561 483 597 391 562 342 407
 490 468 488 457 365 357 463 388 413 351 462 440 493 507 338 377 428 525
 390 473 359 423 519 453 522 429 450 383 469 402 354 422 458 384 656 461
 427 478 523 719 538 515 394 472 433 387 540 530 498 512 452 311 470 541
 331 570 5

>Comprobamos la cantidad de filas y columnas, el nñumero de clientes que han registrado sus datos de vuelo al menos una vez y con cuantos años de registro contamos. Hacemos una última comprobación para saber si las dos columnas que a simple vista parecen vacñías tienen datos; en efecto sí tienen datos más allá de cero.

In [5]:
df_vuelos_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 30.9 MB


>Con un _info_ podemos ver algunas cosas de la base de datos; no hay nulos, todas las columnas tienen valores, también vemos el tipo de dato de cada columna, todos son int  excepto un float en los puntos acumulados. Las columnas de  años y meses están tambien en float.

In [6]:
df_vuelos_info.insert(1,"Date", pd.to_datetime(df_vuelos_info[["Year", "Month"]].assign(day=1)))
df_vuelos_info.drop(columns=["Year", "Month"], inplace=True)

>Vamos a convertir las columnas de _Year_ y _Month_ en una única columna y de paso la creamos como _datetime_. Como solo se especifica mes y año pondremos por defecto dia 1.

In [7]:
##  Comprobamos que se ha guardado la columna Date y que han sido eliminadas Year y Month.
df_vuelos_info.head()

Loyalty Number       Date  Flights Booked  Flights with Companions  \
0          100018 2017-01-01               3                        0   
1          100102 2017-01-01              10                        4   
2          100140 2017-01-01               6                        0   
3          100214 2017-01-01               0                        0   
4          100272 2017-01-01               0                        0   

   Total Flights  Distance  Points Accumulated  Points Redeemed  \
0              3      1521               152.0                0   
1             14      2030               203.0                0   
2              6      1200               120.0                0   
3              0         0                 0.0                0   
4              0         0                 0.0                0   

   Dollar Cost Points Redeemed  
0                            0  
1                            0  
2                            0  
3                            0  
4                            0

In [8]:
## Comprobamos que la columna Date está en datetime
df_vuelos_info.dtypes

Loyalty Number                          int64
Date                           datetime64[ns]
Flights Booked                          int64
Flights with Companions                 int64
Total Flights                           int64
Distance                                int64
Points Accumulated                    float64
Points Redeemed                         int64
Dollar Cost Points Redeemed             int64
dtype: object

In [9]:
## Sacamos información extra de algunas columnas
print(df_vuelos_info[["Distance", "Flights Booked", "Flights with Companions", "Total Flights", "Points Accumulated", "Points Redeemed", "Dollar Cost Points Redeemed"]].describe().T)

                                count         mean          std  min  25%  \
Distance                     405624.0  1208.880059  1433.155320  0.0  0.0   
Flights Booked               405624.0     4.115052     5.225518  0.0  0.0   
Flights with Companions      405624.0     1.031805     2.076869  0.0  0.0   
Total Flights                405624.0     5.146858     6.521227  0.0  0.0   
Points Accumulated           405624.0   123.692721   146.599831  0.0  0.0   
Points Redeemed              405624.0    30.696872   125.486049  0.0  0.0   
Dollar Cost Points Redeemed  405624.0     2.484503    10.150038  0.0  0.0   

                               50%     75%     max  
Distance                     488.0  2336.0  6293.0  
Flights Booked                 1.0     8.0    21.0  
Flights with Companions        0.0     1.0    11.0  
Total Flights                  1.0    10.0    32.0  
Points Accumulated            50.0   239.0   676.5  
Points Redeemed                0.0     0.0   876.0  
Dollar Cost 

In [10]:
## Revisamos el segundo fichero
df_clientes.head()

Loyalty Number Country          Province       City Postal Code  Gender  \
0          480934  Canada           Ontario    Toronto     M2Z 4K1  Female   
1          549612  Canada           Alberta   Edmonton     T3G 6Y6    Male   
2          429460  Canada  British Columbia  Vancouver     V6E 3D9    Male   
3          608370  Canada           Ontario    Toronto     P1W 1K4    Male   
4          530508  Canada            Quebec       Hull     J8Y 3Z5    Male   

  Education    Salary Marital Status Loyalty Card      CLV Enrollment Type  \
0  Bachelor   83236.0        Married         Star  3839.14        Standard   
1   College       NaN       Divorced         Star  3839.61        Standard   
2   College       NaN         Single         Star  3839.75        Standard   
3   College       NaN         Single         Star  3839.75        Standard   
4  Bachelor  103495.0        Married         Star  3842.79        Standard   

   Enrollment Year  Enrollment Month  Cancellation Year  Cancellation Month  
0             2016                 2                NaN                 NaN  
1             2016                 3                NaN                 NaN  
2             2014                 7             2018.0                 1.0  
3             2013                 2                NaN                 NaN  
4             2014                10                NaN                 NaN

>Podemos ver algunas cosas visualizando las 5 primeras filas; Hay columnas numéricas y categóricas, se conecta con la otra base de datos en la columna _Loyalty Number_ que aparece en ambas con el mismo nombre. Podemos ver ya algunos datos nulos en algunas columnas y que hay una fecha nuevamente dividida en dos columnas de mes y año.

In [11]:
## Primeras comprobaciones:
print(df_clientes.shape)
print("-----------------")
print(df_clientes ["Loyalty Card"].unique())
print("-----------------")
print(df_clientes ["Enrollment Type"].unique())
print("-----------------")
print(df_clientes.duplicated(subset="Loyalty Number").sum)

(16737, 16)
-----------------
['Star' 'Aurora' 'Nova']
-----------------
['Standard' '2018 Promotion']
-----------------
<bound method Series.sum of 0        False
1        False
2        False
3        False
4        False
         ...  
16732    False
16733    False
16734    False
16735    False
16736    False
Length: 16737, dtype: bool>


>Comprobamos el número de filas y columnas, también que en las columnas _Loyalty Card_ y _Enrollment Type_ haya más de un dato y por último si hay algún _Loyalty Number_ repetido (para evitar errores de duplicados, estavariable tiene que ser única  ya que cada fila representa a un cliente único). Comparando los resultados de las dos columnas _Loyalty Number_ de ambas tablas vemos tambiém que en ambas hay el mismo número de clientes (16737 registros no repetidos) por lo que probablemente cada cliente de esta tabla tenga registros de sus vuelos en la primera base de datos.

In [12]:
## Hacemos un info
df_clientes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Loyalty Number      16737 non-null  int64  
 1   Country             16737 non-null  object 
 2   Province            16737 non-null  object 
 3   City                16737 non-null  object 
 4   Postal Code         16737 non-null  object 
 5   Gender              16737 non-null  object 
 6   Education           16737 non-null  object 
 7   Salary              12499 non-null  float64
 8   Marital Status      16737 non-null  object 
 9   Loyalty Card        16737 non-null  object 
 10  CLV                 16737 non-null  float64
 11  Enrollment Type     16737 non-null  object 
 12  Enrollment Year     16737 non-null  int64  
 13  Enrollment Month    16737 non-null  int64  
 14  Cancellation Year   2067 non-null   float64
 15  Cancellation Month  2067 non-null   float64
dtypes: f

>Como vimos al principio hay tanto datos de tipo int como object y un par de floats. hay cuatro columnas de fechas de inscripción y cancelación de la cuenta que están en int y float respectivamente, esto último es innecesario ya que años y meses no necesitan decimales. Procederemos a agrupar estas cuatro columnas en dos y a ponerlas en formato de fecha.

>Una última cosa que podemos observar es los datos nulos que tenemos. Solo hay nulos en la columna _salary_, y en las fechas de cancelación.

In [13]:
## agrupamos columnas de fechas y las cambiamos a datetime:
df_clientes.insert(
    12,
    "Enrollment Date",
    pd.to_datetime(
        df_clientes[["Enrollment Year", "Enrollment Month"]]
        .rename(columns={
            "Enrollment Year": "year",
            "Enrollment Month": "month"
        })
        .assign(day=1)
    )
)
df_clientes.drop(columns=["Enrollment Year", "Enrollment Month"], inplace=True)

In [14]:
df_clientes.insert(
    12,
    "Cancellation Date",
    pd.to_datetime(
        df_clientes[["Cancellation Year", "Cancellation Month"]]
        .rename(columns={
            "Cancellation Year": "year",
            "Cancellation Month": "month"
        })
        .assign(day=1)
    )
)
df_clientes.drop(columns=["Cancellation Year", "Cancellation Month"], inplace=True)

In [15]:
df_clientes.head()

Loyalty Number Country          Province       City Postal Code  Gender  \
0          480934  Canada           Ontario    Toronto     M2Z 4K1  Female   
1          549612  Canada           Alberta   Edmonton     T3G 6Y6    Male   
2          429460  Canada  British Columbia  Vancouver     V6E 3D9    Male   
3          608370  Canada           Ontario    Toronto     P1W 1K4    Male   
4          530508  Canada            Quebec       Hull     J8Y 3Z5    Male   

  Education    Salary Marital Status Loyalty Card      CLV Enrollment Type  \
0  Bachelor   83236.0        Married         Star  3839.14        Standard   
1   College       NaN       Divorced         Star  3839.61        Standard   
2   College       NaN         Single         Star  3839.75        Standard   
3   College       NaN         Single         Star  3839.75        Standard   
4  Bachelor  103495.0        Married         Star  3842.79        Standard   

  Cancellation Date Enrollment Date  
0               NaT      2016-02-01  
1               NaT      2016-03-01  
2        2018-01-01      2014-07-01  
3               NaT      2013-02-01  
4               NaT      2014-10-01

In [21]:
## Comprobamos nulos
df_clientes[["Salary", "Cancellation Date"]].isnull().sum()/df_clientes.shape[0]*100

Salary               25.321145
Cancellation Date    87.650117
dtype: float64